src/sentry/tasks/assemble.py

```diff
@@ -311,14 +326,94 @@ def _create_artifact_bundle(
                     artifact_bundle=artifact_bundle,
                     source_file_type=source_file_type.value,
                 )
+        else:
+            raise AssembleArtifactsError(
+                "uploading a bundle without debug ids or release is prohibited"
+            )
+
+
+def get_with_manifest_fallback(value, manifest_key, manifest, throw=True):
+    if value is None:
+        manifest_value = manifest.get(manifest_key)
+        if manifest_value is None and throw:
+            raise AssembleArtifactsError(f"failed to load {manifest_key} from the manifest")
+
+        return manifest_value
 
-            return True
+    return value
 
-    return False
+
+def handle_assemble_for_release_file(bundle, archive, manifest, organization, version):
+    org_slug = manifest.get("org")
+    if organization.slug != org_slug:
+        raise AssembleArtifactsError("organization does not match uploaded bundle")
+
+    release_name = manifest.get("release")
+    if release_name != version:
+        raise AssembleArtifactsError("release does not match uploaded bundle")
+
+    try:
+        release = Release.objects.get(organization_id=organization.id, version=release_name)
+    except Release.DoesNotExist:
+        raise AssembleArtifactsError("release does not exist")
+
+    dist_name = manifest.get("dist")
+    dist = None
+    if release and dist_name:
+        dist = release.add_dist(dist_name)
+
+    artifact_count = len(manifest.get("files", {}))
+    min_artifact_count = options.get("processing.release-archive-min-files")
+    saved_as_archive = False
+
+    if artifact_count >= min_artifact_count:
+        if release is None:
+            raise AssembleArtifactsError("release does not exist")
```

Comment:
didn’t we already asserting this a couple of lines above?

Reply:
Nope, this assertion is that the actual release is existing in the db. The above's assertion checks whether we have a manifest/request data mismatch.